packages/ui/src/Dialog.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { FC, ReactNode, useMemo } from 'react'
+import { Modal, ModalProps, ModalActionProps } from './Modal'
+import { ButtonKind } from './Button'
+
+import styles from './Dialog.module.scss'
+
+export const DIALOG_AFFIRMATION_DEFAULT = 'Are you sure you wish to proceed?'
+
+type DialogActionProps =
+  | {
+      actionOnConfirm: ModalActionProps['onClick']
+      actionChildren: ModalActionProps['children']
+      actionDangerous?: boolean
+      actionDisabled?: ModalActionProps['disabled']
+      actionDisabledTooltip?: ModalActionProps['disabledTooltip']
+    }
+  | {
+      actionDangerous?: never
+      actionDisabled?: never
+      actionDisabledTooltip?: never
+      actionOnConfirm?: never
+      actionChildren?: never
+    }
```

Comment:
nit: Wouldn;t something simpler work as well?
```suggestion
type DialogActionProps =
  | {
      actionOnConfirm: ModalActionProps['onClick']
      actionChildren: ModalActionProps['children']
      actionDangerous?: boolean
      actionDisabled?: ModalActionProps['disabled']
      actionDisabledTooltip?: ModalActionProps['disabledTooltip']
    }
  | {}
```

Reply:
Hmm, we use this pattern at quite some places, so it would be nice to optimise a bit. Using `{}` is rejected by https://github.com/typescript-eslint/typescript-eslint/blob/v4.4.0/packages/eslint-plugin/docs/rules/ban-types.md though.

WDYT about using some type constructor like this one `type NeverObject<T extends object> = T | {[U in keyof T]?: never }`?

Reply:
I like the idea but I would suggest calling the type something like `WholeOrEmpty`

Reply:
Makes sense to me. 👍